controllers/etcd_copy_backups_controller.go

```diff
@@ -0,0 +1,450 @@
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{
+		RequeueAfter: 30 * time.Second,
+	}, nil
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) delete(ctx context.Context, copyTask *druidv1alpha1.EtcdCopyBackupsTask) (ctrl.Result, error) {
+	logger := r.logger.WithValues("task", kutil.Key(copyTask.Namespace, copyTask.Name).String(), "operation", "delete")
+	logger.Info("Starting operation")
+
+	if copyTask.Status.TaskReference != nil {
+		if err := r.updateEtcdCopyBackupsTaskStatus(ctx, copyTask); err != nil {
+			return ctrl.Result{
+				Requeue: true,
+			}, err
+		}
+		if copyTask.Status.LastOperation != nil && reflect.DeepEqual(copyTask.Status.LastOperation.State, druidv1alpha1.LastOperationStateActive) {
+			return ctrl.Result{
+				RequeueAfter: 30 * time.Second,
+			}, errors.New("The resource can not be deleted while still processing")
+		}
+
+		job := &batchv1.Job{}
+		if err := r.Get(ctx, kutil.Key(copyTask.Namespace, copyTask.Name+workerSuffix), job); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{
+				Requeue: true,
+			}, err
+		}
+		if err := r.removeFinalizerFromObject(ctx, job, logger); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{
+				Requeue: true,
+			}, err
+		}
+		if err := r.Delete(ctx, job); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{
+				Requeue: true,
+			}, err
+		}
+	}
+	if err := r.removeFinalizerFromObject(ctx, copyTask, logger); err != nil {
+		return ctrl.Result{
+			Requeue: true,
+		}, err
+	}
+
+	logger.Info(fmt.Sprintf("Deleted ETCDCopyBackupTask %s/%s successfully.", copyTask.Namespace, copyTask.Name))
+	return ctrl.Result{}, nil
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) updateEtcdCopyTaskStatusLastError(ctx context.Context, copyTask *druidv1alpha1.EtcdCopyBackupsTask, err error) error {
+	return kutil.TryUpdateStatus(ctx, retry.DefaultBackoff, r.Client, copyTask, func() error {
+		copyTask.Status.LastError = pointer.StringPtr(err.Error())
+		copyTask.Status.ObservedGeneration = &copyTask.Generation
+		return nil
+	})
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) updateEtcdCopyBackupsTaskStatus(ctx context.Context, copyTask *druidv1alpha1.EtcdCopyBackupsTask) error {
+	return kutil.TryUpdateStatus(ctx, retry.DefaultBackoff, r.Client, copyTask, func() error {
+		copyTask.Status.TaskReference = &corev1.TypedLocalObjectReference{
+			Name:     copyTask.Name + workerSuffix,
+			Kind:     "Job",
+			APIGroup: pointer.StringPtr(batchv1.SchemeGroupVersion.String()),
+		}
+		lastOperation, err := r.getLastOperation(ctx, copyTask)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				copyTask.Status.TaskReference = nil
+				copyTask.Status.LastOperation = nil
+				return nil
+			} else {
+				copyTask.Status.LastError = pointer.StringPtr(err.Error())
+				return err
+			}
+		}
+		copyTask.Status.LastOperation = lastOperation
+		copyTask.Status.ObservedGeneration = &copyTask.Generation
```

Comment:
I find the logic above somewhat hard to follow, I would propose something similar to:
```suggestion
		lastOperation, err := r.getLastOperation(ctx, copyTask)
		if err != nil {
			if apierrors.IsNotFound(err) {
				copyTask.Status.TaskReference = nil
				copyTask.Status.LastOperation = nil
				copyTask.Status.LastError = nil
			} else {
				copyTask.Status.LastError = pointer.StringPtr(err.Error())
			}
		} else {
			copyTask.Status.TaskReference = &corev1.TypedLocalObjectReference{
				Name:     copyTask.Name + workerSuffix,
				Kind:     "Job",
				APIGroup: pointer.StringPtr(batchv1.SchemeGroupVersion.String()),
			}
			copyTask.Status.LastOperation = lastOperation
		}
		copyTask.Status.ObservedGeneration = &copyTask.Generation
```